### DISCUSSION TITLE
WebFX can now access local files!

### DISCUSSION THREAD

**Discussion by salmonb** [webfx-project/webfx#14] [Announcements]
Several people requested this feature, so we made it!

We also made this [new demo](https://files.webfx.dev) to illustrate what WebFX can now do with local files, and gave more explanation in our [blog](https://blog.webfx.dev/2022/10/19/accessing-local-files/).

<p align="center" >
  <a href="https://files.webfx.dev">
  <img src="https://blog.webfx.dev/2022/10/19/accessing-local-files/FilesDemo.png" />
</a>
</p>

You are very welcome to post here any comment or question about this new feature.

**Comment by gnahraf** [comment]
Nice work!

Is there a way to read a byte stream from the file? I'm hoping to parse binary files in code. Here's a simple javascript function on my landing page at https://crums.io that computes the hash of a chosen file. It's not exactly an example of _parsing_ a binary file, but it does demo the feasibility (I hope).

**Reply by salmonb** [reply]
Not presently as I develop features only on demand, but since you asked, I will investigate and let you know what I can do.

**Reply by salmonb** [reply]
Ok, I added the `Future<byte[]> readAsBytes(File file)`method to `dev.webfx.platform.file.FileReader` (ensure to refresh your SNAPSHOTs by running `mvn compile -U` on your project - this should pull this new version). This method is for simple cases, as it doesn't give you a stream but the whole content of the binary file. It's not designed to work with big files, but it may be good enough for your use case. If it's too simple and you really need to stream big files, let me know.

**Comment by gnahraf** [comment]
A suggestion for your blog post, btw.. Ask the reader to clone the example project right off the bat. I couldn't get it to work by editing my existing webfx.xml ..be a shame if another reader gives up prematurely, I'm saying

**Reply by salmonb** [reply]
Thanks for reporting that. I would prefer not to ask people to clone, but rather make my instructions work, so people can easily integrate this feature in their existing project. It was working during my tests, but maybe that didn't work for you because you had old SNAPSHOT artifacts in your local Maven repo when you invoked `webfx update`. Did the CLI complain about not finding the `dev.webfx.platform.file` package? This would confirm that you had an old SNAPSHOT of the WebFX platform. Maybe I should add an option in the CLI to force a SNAPSHOTs reload.

**Comment by gnahraf** [comment]
I did `webfx update` before doing anything else. I might have messed things up, cuz the `webfx.xml` file that you mention needs updating, there are 5 of those. I wasn't sure which you were referring to, so I tried them all, one-by-one. Then, I tried manually updating the pom files.. didn't go well, so I backed out ;) 

But after cloning the demo project, it became clearer for me what `webfx build` does. If I had to guess, the demo project is a layout you'd expect after a `webfx init` with the new features added (after which you could invoke `build`). I'm a noob with this library, so I might have just gotten confused.

**Reply by salmonb** [reply]
Thanks for this feedback, I updated the blog post to say:

>  add this line in the `<required-libraries>` section of the `webfx.xml` located at your project’s root directory

because yes, I understand your confusion, there are as many webfx.xml as pom.xml (one for each module).

I tested again the instructions (I restarted from scratch from the getting started guide and then followed the blog instructions), and it worked for me. I'm pretty sure your issue was with an old SNAPSHOT version in your local Maven repo. So I will add an option in the CLI to force a refresh of the SNAPSHOTs.
